### PR TITLE
Change the barrier class into a record

### PR DESCRIPTION
**Pull request by daviditen**
Changed the barrier class into a record so that it doesn't require explicit
deletes.

Updated chpldoc comments for a few of the methods.

Documented the 'reset' method in order to expose it.

Updated tests to remove deletes of barriers.

Added another barrier to each SPMD LCALS kernel. While removing barrier
deletes, I realized that a barrier was required both before and after
updating the loop index variable, but previously only put one after the
update.

### CODE REVIEW

**Comment by ronawho** [comment]
Doesn't this really want a record-wrapped class? I thought identity was important for the barrier for cases where a barrier is passed between procs. Am I really off-base here?

**Comment by daviditen** [comment]
You're right.  It was previously a class-wrapped class, and is now a record-wrapped class.

**Comment by ronawho** [comment]
Oh duh. Sorry, I didn't actually look at the code and I read the title as "plain old class is now a plain old record"

**Comment by mppf** [comment]
hi @daviditen - this PR may well be a good first step, but it can't possibly be complete.

It looks to me that *before this PR* the class BarrierBaseType instance in the *bar* field is never 'deleted'. That's a memory leak, isn't it? Shouldn't the destructor for Barrier delete this field?

If you add such a destructor to the record version of Barrier, you'll also have to think about whether or not a Barrier is ever copied (e.g. passed by in intent, copied to another variable). If it is, you'll need to also override an initCopy method and possibly a proc = function in order to control the behavior.

I can see several memory management strategies here:
 - reference counting
 - one copy is the "owner" and the other copies do nothing when they are destroyed
 - somehow prohibit copies of this type (e.g. with 'halt' in initCopy / proc = ).

It is not immediately obvious to me which strategy makes sense in this situation.

**Comment by mppf** [comment]
One more thing - this module is called Barrier and it includes a class called Barrier. It's not so clear to me what the implications of that are, but my understanding is that it's bad style if nothing else (since Chapel is a *single namespace* language).

**Comment by noakesmichael** [comment]
Given that Chapel is a single namespace language this is an error and not merely
“bad style”.  As far as I can tell, it’s a bug, my guess is in scope-resolve,  that this
doesn’t generate an error.

Mike



On Dec 7, 2016, at 5:52 AM, Michael Ferguson <notifications@github.com<mailto:notifications@github.com>> wrote:


One more thing - this module is called Barrier and it includes a class called Barrier. It's not so clear to me what the implications of that are, but my understanding is that it's bad style if nothing else (since Chapel is a single namespace language).

**Comment by lydia-duncan** [comment]
Man, I thought I caught all those cases . . .

**Comment by daviditen** [comment]
The Barrier record (previously class) has a destructor, so I don't think it leaks the bar field:

```
pragma "no doc"
proc ~Barrier() {
  if bar != nil {
    delete bar;
  }
}
```

None of the tests using Barriers that currently exist run into a problem relating to copies, so I'll have to create a new test for that.

For the module name vs. type name issue, how about renaming the module to TaskBarrier?

**Comment by mppf** [comment]
Hi @daviditen - yes I see the ~Barrier now (and I don't know why I missed it before). So, what remains to be done is to figure out what happens when you copy a Barrier. With the change from class -> record, it will now do a double-free unless you take some action.

**Comment by daviditen** [comment]
Hi @mppf 

I added an "owned" field that is treated similarly to what happens for the string record.  This means that copies will be shallow and have the same underlying class.  Does that seem reasonable to you?

**Comment by mppf** [comment]
hi @daviditen - yes, that seems good to me. You'll want to note though that it's on the user if a Barrier goes out of scope before a copy of it does.

**Comment by daviditen** [comment]
I added a note to the top-level comment about how barriers get deleted when they go out of scope and any copies are invalid after that.

I'd like to leave renaming the module for another PR.  Does it look good now @mppf ?

File: modules/standard/Barrier.chpl [hunk 1]
```diff
@@ -41,6 +41,10 @@
        writeln("Task ", tid, " is past the barrier");
      }
 
+   Note: When a barrier instance goes out of scope it is automatically deleted.
+   After it is deleted, any copies of the barrier that remain are invalid.
+   Using a copy of a barrier after it has been deleted is undefined.
```
Comment (mppf) [inline modules/standard/Barrier.chpl]: I'd remove this last line entirely, but if not, I'd say "causes undefined behavior" or "is likely to cause a crash" or something slightly more grammatical and hopefully more obvious.

**Comment by daviditen** [comment]
Thanks!

**Pull request merged**